// @ariakit/react-core/cjs/__chunks/6XUE2BI4.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _RYHFE3BIcjs = require('./RYHFE3BI.cjs');


var _5VOPA6D3cjs = require('./5VOPA6D3.cjs');



var _CWY2PHPRcjs = require('./CWY2PHPR.cjs');


var _MPGRRS75cjs = require('./MPGRRS75.cjs');


var _5WLTMVVQcjs = require('./5WLTMVVQ.cjs');




var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/composite/composite-renderer.tsx
var _react = require('react');
function getItemObject(item) {
  if (!item || typeof item !== "object") {
    return { value: item };
  }
  return item;
}
function countItems(items) {
  if (!items)
    return [0];
  if (typeof items === "number") {
    return Array.from({ length: items }, (_, index) => index + 1);
  }
  return items.reduce((count, item, index) => {
    var _a, _b;
    const object = getItemObject(item);
    if (!object.items) {
    }
    if (!object.items) {
      count[index] = index + 1;
      return count;
    }
    const prevCount = (_a = count[index - 1]) != null ? _a : 0;
    const itemsCount = (_b = countItems(object.items)[object.items.length - 1]) != null ? _b : 0;
    count[index] = prevCount + itemsCount;
    return count;
  }, []);
}
function findFirst(items, offset = 1) {
  for (let index = offset > 0 ? 0 : items.length - 1; index >= 0 && index < items.length; index += offset) {
    const item = items[index];
    const object = getItemObject(item);
    if (object.items && findFirst(object.items, offset) !== -1)
      return index;
    if (!object.disabled)
      return index;
  }
  return -1;
}
function findLast(items) {
  return findFirst(items, -1);
}
function findById(items, id, baseId) {
  return items.findIndex((item, index) => {
    var _a;
    const itemId = _RYHFE3BIcjs.getCollectionRendererItemId.call(void 0, item, index, baseId);
    if (itemId === id)
      return true;
    const object = getItemObject(item);
    if ((_a = object.items) == null ? void 0 : _a.length)
      return findById(object.items, id, itemId) !== -1;
    const ids = id.split("/");
    if (ids.length === 1)
      return false;
    return ids.some((id2) => itemId === id2);
  });
}
function useCompositeRenderer(_a) {
  var _b = _a, {
    store,
    orientation: orientationProp,
    persistentIndices: persistentIndicesProp,
    children: renderItem,
    "aria-setsize": ariaSetSize,
    "aria-posinset": ariaPosInSet = 1
  } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, [
    "store",
    "orientation",
    "persistentIndices",
    "children",
    "aria-setsize",
    "aria-posinset"
  ]);
  var _a2;
  const context = _react.useContext.call(void 0, _5VOPA6D3cjs.CompositeContext);
  store = store || context;
  const orientation = (_a2 = _MPGRRS75cjs.useStoreState.call(void 0, 
    store,
    (state) => (orientationProp != null ? orientationProp : state.orientation === "both") ? "vertical" : state.orientation
  )) != null ? _a2 : orientationProp;
  const items = _MPGRRS75cjs.useStoreState.call(void 0, store, (state) => {
    var _a3;
    if ("mounted" in state && state.mounted)
      return 0;
    return (_a3 = props.items) != null ? _a3 : state.items;
  }) || props.items;
  const id = _5WLTMVVQcjs.useId.call(void 0, props.id);
  const itemsCount = _react.useMemo.call(void 0, () => countItems(items), [items]);
  const setSize = _react.useMemo.call(void 0, 
    () => {
      var _a3;
      return (_a3 = ariaSetSize != null ? ariaSetSize : itemsCount[itemsCount.length - 1]) != null ? _a3 : 0;
    },
    [ariaSetSize, itemsCount]
  );
  const firstIndex = _react.useMemo.call(void 0, () => {
    if (!items)
      return -1;
    if (typeof items === "number")
      return 0;
    if (!items.length)
      return -1;
    return findFirst(items);
  }, [items]);
  const lastIndex = _react.useMemo.call(void 0, () => {
    if (!items)
      return -1;
    if (typeof items === "number")
      return items - 1;
    if (!items.length)
      return -1;
    return findLast(items);
  }, [items]);
  const activeId = _MPGRRS75cjs.useStoreState.call(void 0, store, "activeId");
  const activeIndex = _react.useMemo.call(void 0, () => {
    if (!id)
      return -1;
    if (!items)
      return -1;
    if (activeId == null)
      return -1;
    if (typeof items === "number")
      return -1;
    if (!items.length)
      return -1;
    return findById(items, activeId, id);
  }, [id, items, activeId]);
  const persistentIndices = _react.useMemo.call(void 0, () => {
    const indices = [firstIndex, activeIndex, lastIndex].filter(
      (index) => index >= 0
    );
    if (persistentIndicesProp) {
      return [...persistentIndicesProp, ...indices];
    }
    return indices;
  }, [firstIndex, activeIndex, lastIndex, persistentIndicesProp]);
  return _RYHFE3BIcjs.useCollectionRenderer.call(void 0, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {
    id,
    store,
    orientation,
    persistentIndices
  }, props), {
    children: (item) => {
      var _a3;
      const nextItem = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, item), {
        "aria-setsize": setSize,
        "aria-posinset": ariaPosInSet + ((_a3 = itemsCount[item.index - 1]) != null ? _a3 : 0)
      });
      return renderItem == null ? void 0 : renderItem(nextItem);
    }
  }));
}
var CompositeRenderer = _CWY2PHPRcjs.forwardRef.call(void 0, function CompositeRenderer2(props) {
  const htmlProps = useCompositeRenderer(props);
  return _CWY2PHPRcjs.createElement.call(void 0, "div", htmlProps);
});




exports.useCompositeRenderer = useCompositeRenderer; exports.CompositeRenderer = CompositeRenderer;
