// @ariakit/react-core/cjs/__chunks/CWY2PHPR.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } }

var _5WLTMVVQcjs = require('./5WLTMVVQ.cjs');



var _7WUTS26Ycjs = require('./7WUTS26Y.cjs');




var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/utils/system.tsx
var _react = require('react'); var React = _interopRequireWildcard(_react);
var _misc = require('@ariakit/core/utils/misc');
var _jsxruntime = require('react/jsx-runtime');
function isRenderProp(children) {
  return typeof children === "function";
}
function forwardRef2(render) {
  const Role = React.forwardRef((props, ref) => render(_72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), { ref })));
  Role.displayName = render.displayName || render.name;
  return Role;
}
function memo2(Component, propsAreEqual) {
  const Role = React.memo(Component, propsAreEqual);
  Role.displayName = Component.displayName || Component.name;
  return Role;
}
function createComponent(render) {
  const Role = (props, ref) => render(_72I2GWXFcjs.__spreadValues.call(void 0, { ref }, props));
  return React.forwardRef(Role);
}
function createMemoComponent(render) {
  const Role = createComponent(render);
  return React.memo(Role);
}
function createElement(Type, props) {
  const _a = props, { as: As, wrapElement, render } = _a, rest = _72I2GWXFcjs.__objRest.call(void 0, _a, ["as", "wrapElement", "render"]);
  let element;
  const mergedRef = _5WLTMVVQcjs.useMergeRefs.call(void 0, props.ref, _7WUTS26Ycjs.getRefProperty.call(void 0, render));
  if (As && typeof As !== "string") {
    element = /* @__PURE__ */ _jsxruntime.jsx.call(void 0, As, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, rest), { render }));
  } else if (React.isValidElement(render)) {
    const renderProps = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, render.props), { ref: mergedRef });
    element = React.cloneElement(render, _7WUTS26Ycjs.mergeProps.call(void 0, rest, renderProps));
  } else if (render) {
    element = render(rest);
  } else if (isRenderProp(props.children)) {
    const _b = rest, { children } = _b, otherProps = _72I2GWXFcjs.__objRest.call(void 0, _b, ["children"]);
    element = props.children(otherProps);
  } else if (As) {
    element = /* @__PURE__ */ _jsxruntime.jsx.call(void 0, As, _72I2GWXFcjs.__spreadValues.call(void 0, {}, rest));
  } else {
    element = /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Type, _72I2GWXFcjs.__spreadValues.call(void 0, {}, rest));
  }
  if (wrapElement) {
    return wrapElement(element);
  }
  return element;
}
function createHook(useProps) {
  const useRole = (props = {}) => {
    const htmlProps = useProps(props);
    const copy = {};
    for (const prop in htmlProps) {
      if (_misc.hasOwnProperty.call(void 0, htmlProps, prop) && htmlProps[prop] !== void 0) {
        copy[prop] = htmlProps[prop];
      }
    }
    return copy;
  };
  return useRole;
}








exports.forwardRef = forwardRef2; exports.memo = memo2; exports.createComponent = createComponent; exports.createMemoComponent = createMemoComponent; exports.createElement = createElement; exports.createHook = createHook;
