// @ariakit/react-core/cjs/__chunks/DZGTHAPX.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});


var _MPFAD3UKcjs = require('./MPFAD3UK.cjs');


var _Q5EWS32Zcjs = require('./Q5EWS32Z.cjs');


var _A764D2YKcjs = require('./A764D2YK.cjs');


var _SZTCN3SGcjs = require('./SZTCN3SG.cjs');




var _CWY2PHPRcjs = require('./CWY2PHPR.cjs');


var _MPGRRS75cjs = require('./MPGRRS75.cjs');





var _5WLTMVVQcjs = require('./5WLTMVVQ.cjs');




var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/menu/menu-list.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
function useAriaLabelledBy(_a) {
  var _b = _a, { store } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, ["store"]);
  const [id, setId] = _react.useState.call(void 0, void 0);
  const label = props["aria-label"];
  const disclosureElement = store.useState("disclosureElement");
  const contentElement = store.useState("contentElement");
  _react.useEffect.call(void 0, () => {
    const disclosure = disclosureElement;
    if (!disclosure)
      return;
    const menu = contentElement;
    if (!menu)
      return;
    const menuLabel = label || menu.hasAttribute("aria-label");
    if (menuLabel) {
      setId(void 0);
    } else if (disclosure.id) {
      setId(disclosure.id);
    }
  }, [label, disclosureElement, contentElement]);
  return id;
}
var useMenuList = _CWY2PHPRcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, { store, alwaysVisible, composite = true } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, ["store", "alwaysVisible", "composite"]);
    const parentMenu = _react.useContext.call(void 0, _MPFAD3UKcjs.MenuContext);
    const parentMenuBar = _react.useContext.call(void 0, _MPFAD3UKcjs.MenuBarContext);
    const hasParentMenu = !!parentMenu;
    const id = _5WLTMVVQcjs.useId.call(void 0, props.id);
    const onKeyDownProp = props.onKeyDown;
    const dir = store.useState(
      (state) => state.placement.split("-")[0]
    );
    const orientation = store.useState(
      (state) => state.orientation === "both" ? void 0 : state.orientation
    );
    const isHorizontal = orientation !== "vertical";
    const isMenuBarHorizontal = !!_MPGRRS75cjs.useStoreState.call(void 0, 
      parentMenuBar,
      (state) => state.orientation !== "vertical"
    );
    const onKeyDown = _5WLTMVVQcjs.useEvent.call(void 0, (event) => {
      onKeyDownProp == null ? void 0 : onKeyDownProp(event);
      if (event.defaultPrevented)
        return;
      if (hasParentMenu || parentMenuBar && !isHorizontal) {
        const hideMap = {
          ArrowRight: () => dir === "left" && !isHorizontal,
          ArrowLeft: () => dir === "right" && !isHorizontal,
          ArrowUp: () => dir === "bottom" && isHorizontal,
          ArrowDown: () => dir === "top" && isHorizontal
        };
        const action = hideMap[event.key];
        if (action == null ? void 0 : action()) {
          event.stopPropagation();
          event.preventDefault();
          return store.hide();
        }
      }
      if (parentMenuBar) {
        const keyMap = {
          ArrowRight: () => {
            if (!isMenuBarHorizontal)
              return;
            return parentMenuBar.next();
          },
          ArrowLeft: () => {
            if (!isMenuBarHorizontal)
              return;
            return parentMenuBar.previous();
          },
          ArrowDown: () => {
            if (isMenuBarHorizontal)
              return;
            return parentMenuBar.next();
          },
          ArrowUp: () => {
            if (isMenuBarHorizontal)
              return;
            return parentMenuBar.previous();
          }
        };
        const action = keyMap[event.key];
        const id2 = action == null ? void 0 : action();
        if (id2 !== void 0) {
          event.stopPropagation();
          event.preventDefault();
          parentMenuBar.move(id2);
        }
      }
    });
    props = _5WLTMVVQcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _MPFAD3UKcjs.MenuContext.Provider, { value: store, children: element }),
      [store]
    );
    const ariaLabelledBy = useAriaLabelledBy(_72I2GWXFcjs.__spreadValues.call(void 0, { store }, props));
    const mounted = store.useState("mounted");
    const hidden = _SZTCN3SGcjs.isHidden.call(void 0, mounted, props.hidden, alwaysVisible);
    const style = hidden ? _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props.style), { display: "none" }) : props.style;
    props = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {
      id,
      "aria-labelledby": ariaLabelledBy,
      hidden
    }, props), {
      ref: _5WLTMVVQcjs.useMergeRefs.call(void 0, id ? store.setContentElement : null, props.ref),
      style,
      onKeyDown
    });
    if (composite) {
      props = _72I2GWXFcjs.__spreadValues.call(void 0, {
        role: "menu",
        "aria-orientation": orientation
      }, props);
    }
    props = _A764D2YKcjs.useComposite.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, { store, composite }, props));
    props = _Q5EWS32Zcjs.useCompositeTypeahead.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, { store }, props));
    return props;
  }
);
var MenuList = _CWY2PHPRcjs.createComponent.call(void 0, (props) => {
  const htmlProps = useMenuList(props);
  return _CWY2PHPRcjs.createElement.call(void 0, "div", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  MenuList.displayName = "MenuList";
}




exports.useMenuList = useMenuList; exports.MenuList = MenuList;
