// @ariakit/react-core/cjs/__chunks/EHQTSSAA.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _7JKSDWVFcjs = require('./7JKSDWVF.cjs');


var _ZK5KHSMPcjs = require('./ZK5KHSMP.cjs');




var _CWY2PHPRcjs = require('./CWY2PHPR.cjs');


var _MPGRRS75cjs = require('./MPGRRS75.cjs');





var _5WLTMVVQcjs = require('./5WLTMVVQ.cjs');




var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/radio/radio.ts
var _react = require('react');
function getIsChecked(value, storeValue) {
  if (storeValue === void 0)
    return;
  if (value != null && storeValue != null) {
    return storeValue === value;
  }
  return !!storeValue;
}
function isNativeRadio(tagName, type) {
  return tagName === "input" && (!type || type === "radio");
}
var useRadio = _CWY2PHPRcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, { store, value, checked } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, ["store", "value", "checked"]);
    var _a2;
    const context = _react.useContext.call(void 0, _7JKSDWVFcjs.RadioContext);
    store = store || context;
    const id = _5WLTMVVQcjs.useId.call(void 0, props.id);
    const ref = _react.useRef.call(void 0, null);
    const isChecked = (_a2 = _MPGRRS75cjs.useStoreState.call(void 0, 
      store,
      (state) => checked != null ? checked : getIsChecked(value, state.value)
    )) != null ? _a2 : checked;
    _react.useEffect.call(void 0, () => {
      if (!id)
        return;
      if (!isChecked)
        return;
      const isActiveItem = (store == null ? void 0 : store.getState().activeId) === id;
      if (isActiveItem)
        return;
      store == null ? void 0 : store.setActiveId(id);
    }, [store, isChecked, id]);
    const onChangeProp = props.onChange;
    const tagName = _5WLTMVVQcjs.useTagName.call(void 0, ref, props.as || "input");
    const nativeRadio = isNativeRadio(tagName, props.type);
    const onChange = _5WLTMVVQcjs.useEvent.call(void 0, (event) => {
      if (props.disabled) {
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      if (!nativeRadio) {
        event.currentTarget.checked = true;
      }
      onChangeProp == null ? void 0 : onChangeProp(event);
      if (event.defaultPrevented)
        return;
      store == null ? void 0 : store.setValue(value);
    });
    const onClickProp = props.onClick;
    const onClick = _5WLTMVVQcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented)
        return;
      if (nativeRadio)
        return;
      onChange(event);
    });
    const onFocusProp = props.onFocus;
    const onFocus = _5WLTMVVQcjs.useEvent.call(void 0, (event) => {
      onFocusProp == null ? void 0 : onFocusProp(event);
      if (event.defaultPrevented)
        return;
      if (!nativeRadio)
        return;
      if (!store)
        return;
      const { moves, activeId } = store.getState();
      if (!moves)
        return;
      if (id && activeId !== id)
        return;
      onChange(event);
    });
    props = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {
      id,
      role: !nativeRadio ? "radio" : void 0,
      type: nativeRadio ? "radio" : void 0,
      "aria-checked": isChecked
    }, props), {
      ref: _5WLTMVVQcjs.useMergeRefs.call(void 0, ref, props.ref),
      onChange,
      onClick,
      onFocus
    });
    props = _ZK5KHSMPcjs.useCompositeItem.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, { store, clickOnEnter: !nativeRadio }, props));
    return _72I2GWXFcjs.__spreadValues.call(void 0, {
      value: nativeRadio ? value : void 0,
      checked: isChecked
    }, props);
  }
);
var Radio = _CWY2PHPRcjs.createMemoComponent.call(void 0, (props) => {
  const htmlProps = useRadio(props);
  return _CWY2PHPRcjs.createElement.call(void 0, "input", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  Radio.displayName = "Radio";
}




exports.useRadio = useRadio; exports.Radio = Radio;
