// @ariakit/react-core/cjs/__chunks/H76HM4AM.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _NXDHPYNQcjs = require('./NXDHPYNQ.cjs');


var _N2VCZZLPcjs = require('./N2VCZZLP.cjs');




var _CWY2PHPRcjs = require('./CWY2PHPR.cjs');


var _5WLTMVVQcjs = require('./5WLTMVVQ.cjs');




var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/dialog/dialog-dismiss.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var useDialogDismiss = _CWY2PHPRcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, { store } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, ["store"]);
    const context = _react.useContext.call(void 0, _N2VCZZLPcjs.DialogContext);
    store = store || context;
    const onClickProp = props.onClick;
    const onClick = _5WLTMVVQcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented)
        return;
      store == null ? void 0 : store.hide();
    });
    const children = _react.useMemo.call(void 0, 
      () => /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, 
        "svg",
        {
          "aria-label": "Dismiss popup",
          display: "block",
          fill: "none",
          stroke: "currentColor",
          strokeLinecap: "round",
          strokeLinejoin: "round",
          strokeWidth: "1.5pt",
          viewBox: "0 0 16 16",
          height: "1em",
          width: "1em",
          children: [
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "4", x2: "12", y2: "12" }),
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "12", x2: "12", y2: "4" })
          ]
        }
      ),
      []
    );
    props = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {
      "data-dialog-dismiss": "",
      children
    }, props), {
      onClick
    });
    props = _NXDHPYNQcjs.useButton.call(void 0, props);
    return props;
  }
);
var DialogDismiss = _CWY2PHPRcjs.createComponent.call(void 0, (props) => {
  const htmlProps = useDialogDismiss(props);
  return _CWY2PHPRcjs.createElement.call(void 0, "button", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  DialogDismiss.displayName = "DialogDismiss";
}




exports.useDialogDismiss = useDialogDismiss; exports.DialogDismiss = DialogDismiss;
