// @ariakit/react-core/cjs/__chunks/QR3ATHWU.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});



var _CWY2PHPRcjs = require('./CWY2PHPR.cjs');



var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/visually-hidden/visually-hidden.ts
var useVisuallyHidden = _CWY2PHPRcjs.createHook.call(void 0, (props) => {
  props = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), {
    style: _72I2GWXFcjs.__spreadValues.call(void 0, {
      border: 0,
      clip: "rect(0 0 0 0)",
      height: "1px",
      margin: "-1px",
      overflow: "hidden",
      padding: 0,
      position: "absolute",
      whiteSpace: "nowrap",
      width: "1px"
    }, props.style)
  });
  return props;
});
var VisuallyHidden = _CWY2PHPRcjs.createComponent.call(void 0, 
  (props) => {
    const htmlProps = useVisuallyHidden(props);
    return _CWY2PHPRcjs.createElement.call(void 0, "span", htmlProps);
  }
);
if (process.env.NODE_ENV !== "production") {
  VisuallyHidden.displayName = "VisuallyHidden";
}




exports.useVisuallyHidden = useVisuallyHidden; exports.VisuallyHidden = VisuallyHidden;
