// @ariakit/react-core/cjs/__chunks/RXCUKHX3.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _QR3ATHWUcjs = require('./QR3ATHWU.cjs');




var _CWY2PHPRcjs = require('./CWY2PHPR.cjs');



var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/focus-trap/focus-trap.ts
var useFocusTrap = _CWY2PHPRcjs.createHook.call(void 0, (props) => {
  props = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {
    "data-focus-trap": "",
    tabIndex: 0,
    "aria-hidden": true
  }, props), {
    style: _72I2GWXFcjs.__spreadValues.call(void 0, {
      // Prevents unintended scroll jumps.
      position: "fixed",
      top: 0,
      left: 0
    }, props.style)
  });
  props = _QR3ATHWUcjs.useVisuallyHidden.call(void 0, props);
  return props;
});
var FocusTrap = _CWY2PHPRcjs.createComponent.call(void 0, (props) => {
  const htmlProps = useFocusTrap(props);
  return _CWY2PHPRcjs.createElement.call(void 0, "span", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  FocusTrap.displayName = "FocusTrap";
}




exports.useFocusTrap = useFocusTrap; exports.FocusTrap = FocusTrap;
