// @ariakit/react-core/cjs/__chunks/T23EIXNF.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _FO6MB5HFcjs = require('./FO6MB5HF.cjs');




var _CWY2PHPRcjs = require('./CWY2PHPR.cjs');




var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/checkbox/checkbox-check.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var checkmark = /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
  "svg",
  {
    display: "block",
    fill: "none",
    stroke: "currentColor",
    strokeLinecap: "round",
    strokeLinejoin: "round",
    strokeWidth: "1.5pt",
    viewBox: "0 0 16 16",
    height: "1em",
    width: "1em",
    children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "polyline", { points: "4,8 7,12 12,4" })
  }
);
function getChildren(props) {
  if (props.checked) {
    return props.children || checkmark;
  }
  if (typeof props.children === "function") {
    return props.children;
  }
  return null;
}
var useCheckboxCheck = _CWY2PHPRcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, { store, checked } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, ["store", "checked"]);
    const context = _react.useContext.call(void 0, _FO6MB5HFcjs.CheckboxCheckedContext);
    checked = checked != null ? checked : context;
    const children = getChildren({ checked, children: props.children });
    props = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {
      "aria-hidden": true
    }, props), {
      children,
      style: _72I2GWXFcjs.__spreadValues.call(void 0, {
        width: "1em",
        height: "1em",
        pointerEvents: "none"
      }, props.style)
    });
    return props;
  }
);
var CheckboxCheck = _CWY2PHPRcjs.createComponent.call(void 0, (props) => {
  const htmlProps = useCheckboxCheck(props);
  return _CWY2PHPRcjs.createElement.call(void 0, "span", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  CheckboxCheck.displayName = "CheckboxCheck";
}




exports.useCheckboxCheck = useCheckboxCheck; exports.CheckboxCheck = CheckboxCheck;
