// @ariakit/react-core/cjs/__chunks/TUSNAJ32.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _HTEAWNBYcjs = require('./HTEAWNBY.cjs');


var _AGPJKWIOcjs = require('./AGPJKWIO.cjs');


var _SZTCN3SGcjs = require('./SZTCN3SG.cjs');


var _OVOYNBFCcjs = require('./OVOYNBFC.cjs');



var _5WLTMVVQcjs = require('./5WLTMVVQ.cjs');



var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/dialog/dialog-backdrop.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
function DialogBackdrop({
  store,
  backdrop,
  backdropProps,
  alwaysVisible,
  hidden
}) {
  const ref = _react.useRef.call(void 0, null);
  const disclosure = _OVOYNBFCcjs.useDisclosureStore.call(void 0, { disclosure: store });
  const contentElement = store.useState("contentElement");
  _5WLTMVVQcjs.useSafeLayoutEffect.call(void 0, () => {
    const backdrop2 = ref.current;
    const dialog = contentElement;
    if (!backdrop2)
      return;
    if (!dialog)
      return;
    backdrop2.style.zIndex = getComputedStyle(dialog).zIndex;
  }, [contentElement]);
  _5WLTMVVQcjs.useSafeLayoutEffect.call(void 0, () => {
    const id = contentElement == null ? void 0 : contentElement.id;
    if (!id)
      return;
    const backdrop2 = ref.current;
    if (!backdrop2)
      return;
    return _HTEAWNBYcjs.markAncestor.call(void 0, backdrop2, id);
  }, [contentElement]);
  if (hidden != null) {
    backdropProps = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, backdropProps), { hidden });
  }
  const props = _SZTCN3SGcjs.useDisclosureContent.call(void 0, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {
    store: disclosure,
    role: "presentation",
    "data-backdrop": (contentElement == null ? void 0 : contentElement.id) || "",
    alwaysVisible
  }, backdropProps), {
    ref: _5WLTMVVQcjs.useMergeRefs.call(void 0, backdropProps == null ? void 0 : backdropProps.ref, ref),
    style: _72I2GWXFcjs.__spreadValues.call(void 0, {
      position: "fixed",
      top: 0,
      right: 0,
      bottom: 0,
      left: 0
    }, backdropProps == null ? void 0 : backdropProps.style)
  }));
  if (!backdrop)
    return null;
  if (_react.isValidElement.call(void 0, backdrop)) {
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _AGPJKWIOcjs.Role, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), { render: backdrop }));
  }
  const Component = typeof backdrop !== "boolean" ? backdrop : "div";
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _AGPJKWIOcjs.Role, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), { render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Component, {}) }));
}



exports.DialogBackdrop = DialogBackdrop;
