// @ariakit/core/cjs/__chunks/DPHJ32KR.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});

var _PIEJYPN7cjs = require('./PIEJYPN7.cjs');




var _FS2K2N7Ucjs = require('./FS2K2N7U.cjs');


var _RCYVACJOcjs = require('./RCYVACJO.cjs');




var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/popover/popover-store.ts
function createPopoverStore(_a = {}) {
  var _b = _a, {
    popover: otherPopover
  } = _b, props = _72I2GWXFcjs.__objRest.call(void 0, _b, [
    "popover"
  ]);
  const store = _FS2K2N7Ucjs.mergeStore.call(void 0, 
    props.store,
    _FS2K2N7Ucjs.omit.call(void 0, otherPopover, [
      "arrowElement",
      "anchorElement",
      "contentElement",
      "popoverElement",
      "disclosureElement"
    ])
  );
  const syncState = store == null ? void 0 : store.getState();
  const dialog = _PIEJYPN7cjs.createDialogStore.call(void 0, _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), { store }));
  const placement = _RCYVACJOcjs.defaultValue.call(void 0, 
    props.placement,
    syncState == null ? void 0 : syncState.placement,
    "bottom"
  );
  const initialState = _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, dialog.getState()), {
    placement,
    currentPlacement: placement,
    anchorElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.anchorElement, null),
    popoverElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.popoverElement, null),
    arrowElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.arrowElement, null),
    rendered: Symbol("rendered")
  });
  const popover = _FS2K2N7Ucjs.createStore.call(void 0, initialState, dialog, store);
  return _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, dialog), popover), {
    setAnchorElement: (element) => popover.setState("anchorElement", element),
    setPopoverElement: (element) => popover.setState("popoverElement", element),
    setArrowElement: (element) => popover.setState("arrowElement", element),
    render: () => popover.setState("rendered", Symbol("rendered"))
  });
}



exports.createPopoverStore = createPopoverStore;
