// @ariakit/core/cjs/__chunks/V7OHZPBO.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});





var _FS2K2N7Ucjs = require('./FS2K2N7U.cjs');


var _RCYVACJOcjs = require('./RCYVACJO.cjs');



var _72I2GWXFcjs = require('./72I2GWXF.cjs');

// src/disclosure/disclosure-store.ts
function createDisclosureStore(props = {}) {
  const store = _FS2K2N7Ucjs.mergeStore.call(void 0, 
    props.store,
    _FS2K2N7Ucjs.omit.call(void 0, props.disclosure, ["contentElement", "disclosureElement"])
  );
  const syncState = store == null ? void 0 : store.getState();
  const open = _RCYVACJOcjs.defaultValue.call(void 0, 
    props.open,
    syncState == null ? void 0 : syncState.open,
    props.defaultOpen,
    false
  );
  const animated = _RCYVACJOcjs.defaultValue.call(void 0, props.animated, syncState == null ? void 0 : syncState.animated, false);
  const initialState = {
    open,
    animated,
    animating: !!animated && open,
    mounted: open,
    contentElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.contentElement, null),
    disclosureElement: _RCYVACJOcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.disclosureElement, null)
  };
  const disclosure = _FS2K2N7Ucjs.createStore.call(void 0, initialState, store);
  _FS2K2N7Ucjs.setup.call(void 0, 
    disclosure,
    () => _FS2K2N7Ucjs.sync.call(void 0, disclosure, ["animated", "animating"], (state) => {
      if (state.animated)
        return;
      disclosure.setState("animating", false);
    })
  );
  _FS2K2N7Ucjs.setup.call(void 0, 
    disclosure,
    () => _FS2K2N7Ucjs.sync.call(void 0, disclosure, ["open", "animated"], (state, prev) => {
      if (!state.animated)
        return;
      const mounting = state === prev;
      const animating = mounting ? state.open : state.open !== prev.open;
      disclosure.setState("animating", animating);
    })
  );
  _FS2K2N7Ucjs.setup.call(void 0, 
    disclosure,
    () => _FS2K2N7Ucjs.sync.call(void 0, disclosure, ["open", "animating"], (state) => {
      disclosure.setState("mounted", state.open || state.animating);
    })
  );
  return _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, disclosure), {
    setOpen: (value) => disclosure.setState("open", value),
    show: () => disclosure.setState("open", true),
    hide: () => disclosure.setState("open", false),
    toggle: () => disclosure.setState("open", (open2) => !open2),
    stopAnimation: () => disclosure.setState("animating", false),
    setContentElement: (value) => disclosure.setState("contentElement", value),
    setDisclosureElement: (value) => disclosure.setState("disclosureElement", value)
  });
}



exports.createDisclosureStore = createDisclosureStore;
