// @ariakit/react-core/cjs/menu/menu-store.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } }


var _MPFAD3UKcjs = require('../__chunks/MPFAD3UK.cjs');



var _CRKIULULcjs = require('../__chunks/CRKIULUL.cjs');
require('../__chunks/ATSMFQCO.cjs');



var _BOBM73LTcjs = require('../__chunks/BOBM73LT.cjs');
require('../__chunks/OJKRJ6N2.cjs');
require('../__chunks/HOHFF2K6.cjs');
require('../__chunks/OVOYNBFC.cjs');




var _MPGRRS75cjs = require('../__chunks/MPGRRS75.cjs');
require('../__chunks/5WLTMVVQ.cjs');
require('../__chunks/7WUTS26Y.cjs');



var _72I2GWXFcjs = require('../__chunks/72I2GWXF.cjs');

// src/menu/menu-store.ts
var _react = require('react');
var _menustore = require('@ariakit/core/menu/menu-store'); var Core = _interopRequireWildcard(_menustore);
function useMenuStoreOptions(props) {
  var _a, _b, _c, _d, _e, _f;
  const state = (_b = (_a = props.store) == null ? void 0 : _a.getState) == null ? void 0 : _b.call(_a);
  const parentMenu = _react.useContext.call(void 0, _MPFAD3UKcjs.MenuContext);
  const parentMenuBar = _react.useContext.call(void 0, _MPFAD3UKcjs.MenuBarContext);
  const placementProp = props.placement;
  const placement = _MPGRRS75cjs.useStoreState.call(void 0, 
    parentMenu || parentMenuBar,
    (state2) => placementProp || (state2.orientation === "vertical" ? "right-start" : "bottom-start")
  );
  const parentIsMenuBar = !!parentMenuBar && !parentMenu;
  const timeout = parentIsMenuBar ? 0 : 150;
  return _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, _CRKIULULcjs.useCompositeStoreOptions.call(void 0, props)), _BOBM73LTcjs.useHovercardStoreOptions.call(void 0, props)), {
    // TODO: Pass parent prop and remove these default values
    values: (_d = (_c = props.values) != null ? _c : state == null ? void 0 : state.values) != null ? _d : props.defaultValues,
    timeout: (_f = (_e = props.timeout) != null ? _e : state == null ? void 0 : state.timeout) != null ? _f : timeout,
    placement
  });
}
function useMenuStoreProps(store, props) {
  const parentMenu = _react.useContext.call(void 0, _MPFAD3UKcjs.MenuContext);
  store = _CRKIULULcjs.useCompositeStoreProps.call(void 0, store, props);
  store = _BOBM73LTcjs.useHovercardStoreProps.call(void 0, store, props);
  _MPGRRS75cjs.useStoreProps.call(void 0, store, props, "values", "setValues");
  return _react.useMemo.call(void 0, 
    () => _72I2GWXFcjs.__spreadProps.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, store), {
      hideAll: () => {
        store.hide();
        parentMenu == null ? void 0 : parentMenu.hideAll();
      }
    }),
    [store]
  );
}
function useMenuStore(props = {}) {
  const options = useMenuStoreOptions(props);
  const store = _MPGRRS75cjs.useStore.call(void 0, () => Core.createMenuStore(_72I2GWXFcjs.__spreadValues.call(void 0, _72I2GWXFcjs.__spreadValues.call(void 0, {}, props), options)));
  return useMenuStoreProps(store, props);
}




exports.useMenuStore = useMenuStore; exports.useMenuStoreOptions = useMenuStoreOptions; exports.useMenuStoreProps = useMenuStoreProps;
