// @wordpress/components/build-module/toggle-group-control/toggle-group-control-option-base/component.js
import { createElement, Fragment } from "@wordpress/element";
/**
 * External dependencies
 */

// eslint-disable-next-line no-restricted-imports
import { Radio } from '@ariakit/react/radio';
// eslint-disable-next-line no-restricted-imports
import { motion, useReducedMotion } from 'framer-motion';

/**
 * WordPress dependencies
 */
import { useInstanceId } from '@wordpress/compose';
import { useMemo } from '@wordpress/element';

/**
 * Internal dependencies
 */

import { contextConnect, useContextSystem } from '../../ui/context';
import { useToggleGroupControlContext } from '../context';
import * as styles from './styles';
import { useCx } from '../../utils/hooks';
import Tooltip from '../../tooltip';
const {
  ButtonContentView,
  LabelView
} = styles;
const REDUCED_MOTION_TRANSITION_CONFIG = {
  duration: 0
};
const LAYOUT_ID = 'toggle-group-backdrop-shared-layout-id';
const WithToolTip = ({
  showTooltip,
  text,
  children
}) => {
  if (showTooltip && text) {
    return createElement(Tooltip, {
      text: text,
      placement: "top"
    }, children);
  }
  return createElement(Fragment, null, children);
};
function ToggleGroupControlOptionBase(props, forwardedRef) {
  const shouldReduceMotion = useReducedMotion();
  const toggleGroupControlContext = useToggleGroupControlContext();
  const id = useInstanceId(ToggleGroupControlOptionBase, toggleGroupControlContext.baseId || 'toggle-group-control-option-base');
  const buttonProps = useContextSystem({
    ...props,
    id
  }, 'ToggleGroupControlOptionBase');
  const {
    isBlock = false,
    isDeselectable = false,
    size = 'default'
  } = toggleGroupControlContext;
  const {
    className,
    isIcon = false,
    value,
    children,
    showTooltip = false,
    onFocus: onFocusProp,
    ...otherButtonProps
  } = buttonProps;
  const isPressed = toggleGroupControlContext.value === value;
  const cx = useCx();
  const labelViewClasses = useMemo(() => cx(isBlock && styles.labelBlock), [cx, isBlock]);
  const itemClasses = useMemo(() => cx(styles.buttonView({
    isDeselectable,
    isIcon,
    isPressed,
    size
  }), className), [cx, isDeselectable, isIcon, isPressed, size, className]);
  const backdropClasses = useMemo(() => cx(styles.backdropView), [cx]);
  const buttonOnClick = () => {
    if (isDeselectable && isPressed) {
      toggleGroupControlContext.setValue(undefined);
    } else {
      toggleGroupControlContext.setValue(value);
    }
  };
  const commonProps = {
    ...otherButtonProps,
    className: itemClasses,
    'data-value': value,
    ref: forwardedRef
  };
  return createElement(LabelView, {
    className: labelViewClasses
  }, createElement(WithToolTip, {
    showTooltip: showTooltip,
    text: otherButtonProps['aria-label']
  }, isDeselectable ? createElement("button", {
    ...commonProps,
    onFocus: onFocusProp,
    "aria-pressed": isPressed,
    type: "button",
    onClick: buttonOnClick
  }, createElement(ButtonContentView, null, children)) : createElement(Radio, {
    render: createElement("button", {
      ...commonProps,
      onFocus: event => {
        onFocusProp?.(event);
        if (event.defaultPrevented) return;
        toggleGroupControlContext.setValue(value);
      }
    }),
    value: value
  }, createElement(ButtonContentView, null, children))), isPressed ? createElement(motion.div, {
    className: backdropClasses,
    transition: shouldReduceMotion ? REDUCED_MOTION_TRANSITION_CONFIG : undefined,
    role: "presentation",
    layoutId: LAYOUT_ID
  }) : null);
}

/**
 * `ToggleGroupControlOptionBase` is a form component and is meant to be used as an internal,
 * generic component for any children of `ToggleGroupControl`.
 *
 * @example
 * ```jsx
 * import {
 *   __experimentalToggleGroupControl as ToggleGroupControl,
 *   __experimentalToggleGroupControlOptionBase as ToggleGroupControlOptionBase,
 * } from '@wordpress/components';
 *
 * function Example() {
 *   return (
 *     <ToggleGroupControl label="my label" value="vertical" isBlock>
 *       <ToggleGroupControlOption value="horizontal" label="Horizontal" />
 *       <ToggleGroupControlOption value="vertical" label="Vertical" />
 *     </ToggleGroupControl>
 *   );
 * }
 * ```
 */
const ConnectedToggleGroupControlOptionBase = contextConnect(ToggleGroupControlOptionBase, 'ToggleGroupControlOptionBase');
export default ConnectedToggleGroupControlOptionBase;
