// @wordpress/components/build/toolbar/toolbar/toolbar-container.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ToolbarContainer = void 0;
var _element = require("@wordpress/element");
var _toolbar = require("@ariakit/react/toolbar");
var _i18n = require("@wordpress/i18n");
var _toolbarContext = _interopRequireDefault(require("../toolbar-context"));
/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

function UnforwardedToolbarContainer({
  label,
  ...props
}, ref) {
  const toolbarStore = (0, _toolbar.useToolbarStore)({
    focusLoop: true,
    rtl: (0, _i18n.isRTL)()
  });
  return (
    // This will provide state for `ToolbarButton`'s
    (0, _element.createElement)(_toolbarContext.default.Provider, {
      value: toolbarStore
    }, (0, _element.createElement)(_toolbar.Toolbar, {
      ref: ref,
      "aria-label": label,
      store: toolbarStore,
      ...props
    }))
  );
}
const ToolbarContainer = (0, _element.forwardRef)(UnforwardedToolbarContainer);
exports.ToolbarContainer = ToolbarContainer;
var _default = ToolbarContainer;
exports.default = _default;
