// @wordpress/components/build-module/spinner/index.js
import { createElement } from "@wordpress/element";
/**
 * External dependencies
 */
import classNames from 'classnames';
/**
 * Internal dependencies
 */
import { StyledSpinner, SpinnerTrack, SpinnerIndicator } from './styles';
/**
 * WordPress dependencies
 */
import { forwardRef } from '@wordpress/element';
export function UnforwardedSpinner({
  className,
  ...props
}, forwardedRef) {
  return createElement(StyledSpinner, {
    className: classNames('components-spinner', className),
    viewBox: "0 0 100 100",
    width: "16",
    height: "16",
    xmlns: "http://www.w3.org/2000/svg",
    role: "presentation",
    focusable: "false",
    ...props,
    ref: forwardedRef
  }, createElement(SpinnerTrack, {
    cx: "50",
    cy: "50",
    r: "50",
    vectorEffect: "non-scaling-stroke"
  }), createElement(SpinnerIndicator, {
    d: "m 50 0 a 50 50 0 0 1 50 50",
    vectorEffect: "non-scaling-stroke"
  }));
}
/**
 * `Spinner` is a component used to notify users that their action is being processed.
 *
 * ```js
 *   import { Spinner } from '@wordpress/components';
 *
 *   function Example() {
 *     return <Spinner />;
 *   }
 * ```
 */
export const Spinner = forwardRef(UnforwardedSpinner);
export default Spinner;
