// @wordpress/components/build/toggle-group-control/toggle-group-control-option-base/component.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _element = require("@wordpress/element");
var _radio = require("@ariakit/react/radio");
var _framerMotion = require("framer-motion");
var _compose = require("@wordpress/compose");
var _context = require("../../ui/context");
var _context2 = require("../context");
var styles = _interopRequireWildcard(require("./styles"));
var _hooks = require("../../utils/hooks");
var _tooltip = _interopRequireDefault(require("../../tooltip"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/**
 * External dependencies
 */

// eslint-disable-next-line no-restricted-imports

// eslint-disable-next-line no-restricted-imports

/**
 * WordPress dependencies
 */

const {
  ButtonContentView,
  LabelView
} = styles;
const REDUCED_MOTION_TRANSITION_CONFIG = {
  duration: 0
};
const LAYOUT_ID = 'toggle-group-backdrop-shared-layout-id';
const WithToolTip = ({
  showTooltip,
  text,
  children
}) => {
  if (showTooltip && text) {
    return (0, _element.createElement)(_tooltip.default, {
      text: text,
      placement: "top"
    }, children);
  }
  return (0, _element.createElement)(_element.Fragment, null, children);
};
function ToggleGroupControlOptionBase(props, forwardedRef) {
  const shouldReduceMotion = (0, _framerMotion.useReducedMotion)();
  const toggleGroupControlContext = (0, _context2.useToggleGroupControlContext)();
  const id = (0, _compose.useInstanceId)(ToggleGroupControlOptionBase, toggleGroupControlContext.baseId || 'toggle-group-control-option-base');
  const buttonProps = (0, _context.useContextSystem)({
    ...props,
    id
  }, 'ToggleGroupControlOptionBase');
  const {
    isBlock = false,
    isDeselectable = false,
    size = 'default'
  } = toggleGroupControlContext;
  const {
    className,
    isIcon = false,
    value,
    children,
    showTooltip = false,
    onFocus: onFocusProp,
    ...otherButtonProps
  } = buttonProps;
  const isPressed = toggleGroupControlContext.value === value;
  const cx = (0, _hooks.useCx)();
  const labelViewClasses = (0, _element.useMemo)(() => cx(isBlock && styles.labelBlock), [cx, isBlock]);
  const itemClasses = (0, _element.useMemo)(() => cx(styles.buttonView({
    isDeselectable,
    isIcon,
    isPressed,
    size
  }), className), [cx, isDeselectable, isIcon, isPressed, size, className]);
  const backdropClasses = (0, _element.useMemo)(() => cx(styles.backdropView), [cx]);
  const buttonOnClick = () => {
    if (isDeselectable && isPressed) {
      toggleGroupControlContext.setValue(undefined);
    } else {
      toggleGroupControlContext.setValue(value);
    }
  };
  const commonProps = {
    ...otherButtonProps,
    className: itemClasses,
    'data-value': value,
    ref: forwardedRef
  };
  return (0, _element.createElement)(LabelView, {
    className: labelViewClasses
  }, (0, _element.createElement)(WithToolTip, {
    showTooltip: showTooltip,
    text: otherButtonProps['aria-label']
  }, isDeselectable ? (0, _element.createElement)("button", {
    ...commonProps,
    onFocus: onFocusProp,
    "aria-pressed": isPressed,
    type: "button",
    onClick: buttonOnClick
  }, (0, _element.createElement)(ButtonContentView, null, children)) : (0, _element.createElement)(_radio.Radio, {
    render: (0, _element.createElement)("button", {
      ...commonProps,
      onFocus: event => {
        onFocusProp?.(event);
        if (event.defaultPrevented) return;
        toggleGroupControlContext.setValue(value);
      }
    }),
    value: value
  }, (0, _element.createElement)(ButtonContentView, null, children))), isPressed ? (0, _element.createElement)(_framerMotion.motion.div, {
    className: backdropClasses,
    transition: shouldReduceMotion ? REDUCED_MOTION_TRANSITION_CONFIG : undefined,
    role: "presentation",
    layoutId: LAYOUT_ID
  }) : null);
}

/**
 * `ToggleGroupControlOptionBase` is a form component and is meant to be used as an internal,
 * generic component for any children of `ToggleGroupControl`.
 *
 * @example
 * ```jsx
 * import {
 *   __experimentalToggleGroupControl as ToggleGroupControl,
 *   __experimentalToggleGroupControlOptionBase as ToggleGroupControlOptionBase,
 * } from '@wordpress/components';
 *
 * function Example() {
 *   return (
 *     <ToggleGroupControl label="my label" value="vertical" isBlock>
 *       <ToggleGroupControlOption value="horizontal" label="Horizontal" />
 *       <ToggleGroupControlOption value="vertical" label="Vertical" />
 *     </ToggleGroupControl>
 *   );
 * }
 * ```
 */
const ConnectedToggleGroupControlOptionBase = (0, _context.contextConnect)(ToggleGroupControlOptionBase, 'ToggleGroupControlOptionBase');
var _default = ConnectedToggleGroupControlOptionBase;
exports.default = _default;
