// @wordpress/components/build/select-control/index.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SelectControl = void 0;
var _element = require("@wordpress/element");
var _pickerCell = _interopRequireDefault(require("../mobile/bottom-sheet/picker-cell"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const SelectControl = (0, _element.memo)(({
  help,
  instanceId,
  label,
  multiple = false,
  onChange,
  options = [],
  className,
  hideLabelFromVision,
  ...props
}) => {
  const id = `inspector-select-control-${instanceId}`;
  return (0, _element.createElement)(_pickerCell.default, {
    label: label,
    hideLabelFromVision: hideLabelFromVision,
    id: id,
    help: help,
    className: className,
    onChangeValue: onChange,
    "aria-describedby": !!help ? `${id}__help` : undefined,
    multiple: multiple,
    options: options,
    ...props
  });
});
exports.SelectControl = SelectControl;
var _default = SelectControl;
exports.default = _default;
